### PR TITLE
LPS-138929 Number of Account Roles is not accurate with the pagination set

### PR DESCRIPTION
**Pull request by erickmonteiroo**

### CODE REVIEW

**Comment by erickmonteiroo** [comment]
ci:test:sf

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java [hunk 1]
```diff
@@ -267,23 +272,38 @@ public boolean hasUserAccountRole(
...
+			companyId, accountEntryIds, keywords, start, end, orderByComparator,
+			null);
+	}
 
-		if (roleDynamicQuery == null) {
+	@Override
+	public BaseModelSearchResult<AccountRole> searchAccountRoles(
+		long companyId, long[] accountEntryIds, String keywords, int start,
+		int end, OrderByComparator<?> orderByComparator,
+		Map<String, Object> params) {
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java:284]: @erickmonteiroo if you do a quick search for other service methods that follow this pattern, you'll see that the `params` map is considered part of the filtering, and always comes before any pagination and sorting parameters (e.g. `start`, `end`, `orderByComparator`).

File: modules/apps/account/account-test/src/testIntegration/java/com/liferay/account/service/test/AccountRoleLocalServiceTest.java [hunk 2]
```diff
@@ -495,6 +497,40 @@ public void testSearchAccountRolesWithPagination() throws Exception {
...
+			HashMapBuilder.<String, Object>put(
+				"excludedRoleIds", new Long[] {accountRole.getRoleId()}
+			).build(),
+			_accountRoleLocalService.searchAccountRoles(
+				accountRole.getCompanyId(),
+				new long[] {accountRole.getAccountEntryId()}, null,
+				QueryUtil.ALL_POS, QueryUtil.ALL_POS, null,
+				HashMapBuilder.<String, Object>put(
+					"excludedRoleIds", new Long[] {accountRole.getRoleId()}
+				).build()));
```
Comment (drewbrokke) [inline modules/apps/account/account-test/src/testIntegration/java/com/liferay/account/service/test/AccountRoleLocalServiceTest.java:531]: @erickmonteiroo for each of these `expectedAccountRoles` parameters, you should be passing in a manually created list instead of search results. Otherwise this test will never fail, because each test case will just compare the results of an operation against the results of an equal operation.

**Review by drewbrokke — CHANGES_REQUESTED** [review]
@erickmonteiroo this is overall looking good, just a couple of adjustments needed:

- [ ] Check service method parameter ordering
- [ ] Fix test case expected values

**Comment by erickmonteiroo** [comment]
ci:test:sf

File: modules/apps/account/account-test/src/testIntegration/java/com/liferay/account/service/test/AccountRoleLocalServiceTest.java (hunk 2 above)
Comment (erickmonteiroo) [reply modules/apps/account/account-test/src/testIntegration/java/com/liferay/account/service/test/AccountRoleLocalServiceTest.java]: Also made some additional changes to the test, such as adding 2 roles to make sure we're hiding the expected role instead of just having an empty list as return.

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by erickmonteiroo** [comment]
ci:stop

**Comment by erickmonteiroo** [comment]
ci:test:relevant

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java [hunk 3]
```diff
@@ -353,53 +373,91 @@ private void _checkAccountRole(Company company, String roleName)
...
+				if (Validator.isNotNull(params.get("excludedRoleNames"))) {
+					predicate = predicate.and(
+						RoleTable.INSTANCE.name.notIn(
+							(String[])params.get("excludedRoleNames")));
+				}
+
+				if (Validator.isNotNull(params.get("excludedRoleIds"))) {
+					predicate = predicate.and(
+						RoleTable.INSTANCE.roleId.notIn(
+							(Long[])params.get("excludedRoleIds")));
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java:430]: Nit: Extract each `param.get` into a variable so we don't call `get` twice for each case.

**Review by drewbrokke — CHANGES_REQUESTED** [review]
@erickmonteiroo we're very close, only small things to fix this time.

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by patriciajeanperez** [comment]
ci:test:account

**Comment by patriciajeanperez** [comment]
ci:test:relevant

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java [hunk 4]
```diff
@@ -267,23 +272,38 @@ public boolean hasUserAccountRole(
...
-			companyId, accountEntryIds, keywords, orderByComparator);
+		return searchAccountRoles(
+			companyId, accountEntryIds, keywords, null, start, end,
+			orderByComparator);
+	}
+
+	@Override
+	public BaseModelSearchResult<AccountRole> searchAccountRoles(
+		long companyId, long[] accountEntryIds, String keywords,
+		Map<String, Object> params, int start, int end,
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java:283]: Ok so I just double-checked with Brian, and it looks like we need to change this to use a `LinkedHashMap`. I'm not sure it matters in this particular case, but there are times when the order of parameters matters.

- [ ] Apply `LinkedHashMap`

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java [hunk 5]
```diff
@@ -353,53 +373,101 @@ private void _checkAccountRole(Company company, String roleName)
 			roleName, null, _roleDescriptionsMaps.get(roleName));
 	}
 
-	private DynamicQuery _getRoleDynamicQuery(
+	private DSLQuery _getDSLQuery(
 		long companyId, long[] accountEntryIds, String keywords,
+		Map<String, Object> params, int start, int end,
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java:378]: - [ ] Apply `LinkedHashMap`

File: modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java [hunk 6]
```diff
@@ -353,53 +373,101 @@ private void _checkAccountRole(Company company, String roleName)
...
-				RestrictionsFactoryUtil.ilike(
-					"title", StringUtil.quote(keywords, StringPool.PERCENT)));
+		return dslQuery;
+	}
 
-			roleDynamicQuery.add(disjunction);
-		}
+	private GroupByStep _getGroupByStep(
+		long[] accountEntryIds, long companyId, FromStep fromStep,
+		String keywords, Map<String, Object> params) {
```
Comment (drewbrokke) [inline modules/apps/account/account-service/src/main/java/com/liferay/account/service/impl/AccountRoleLocalServiceImpl.java:403]: - [ ] Apply `LinkedHashMap`

**Review by drewbrokke — CHANGES_REQUESTED** [review]

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Comment by drewbrokke** [comment]
Sorry @erickmonteiroo looks like a merge conflict here. Please rebase and re-send.

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

File: modules/apps/account/account-test/src/testIntegration/java/com/liferay/account/service/test/AccountRoleLocalServiceTest.java [hunk 7]
```diff
@@ -516,6 +545,21 @@ private void _assertHasPermission(
...
+		LinkedHashMap<String, Object> params,
+		BaseModelSearchResult<AccountRole> expectedAccountRoles) {
+
+		BaseModelSearchResult<AccountRole> accountRoleBaseModelSearchResult =
+			_accountRoleLocalService.searchAccountRoles(
+				companyId, accountEntryIds, null, params, QueryUtil.ALL_POS,
+				QueryUtil.ALL_POS, null);
+
+		Assert.assertEquals(
+			expectedAccountRoles.getBaseModels(),
```
Comment (drewbrokke) [inline modules/apps/account/account-test/src/testIntegration/java/com/liferay/account/service/test/AccountRoleLocalServiceTest.java:559]: Since you're not using the `length` of the expected `BaseModelSearchResult` and only the list of models, just pass in a list of expected models.

**Review by drewbrokke — CHANGES_REQUESTED** [review]
@erickmonteiroo there are a few more refinements we can make to tighten things up. See my inline comments.

**Comment by erickmonteiroo** [comment]
ci:test:sf

**Comment by erickmonteiroo** [comment]
ci:test:relevant

**Review by drewbrokke — APPROVED** [review]
LGTM 👍  Nice work @erickmonteiroo!

**Comment by drewbrokke** [comment]
ci:forward
